### ISSUE TITLE
Records dropped on large files when partition breaks rowTag

### ISSUE THREAD

**Issue by jimenefe** [databricks/spark-xml#398]
We have found an issue where records are dropped silently.

By default, the input XML file will be partitioned into several "splits" with size as set in the 'fs.local.block.size' config setting (by default 32MB). We've found that when this split happens in the middle of the starting row tag element, the record is lost and not read into the DataFrame.

The easy workaround is to set 'fs.local.block.size' larger than the file size, which is guaranteed to work, but has a severe performance penalty for large files.

To check this scenario, I've setup a quick and dirty test as follows (inside `XmlSuite`):

```scala
test("fails on file partitioning") {

    val xml =
      s"""<people>
         |   <person>
         |     <age born="1990-02-24">25</age>
         |     <name>Hyukjin</name>
         |   </person>
         |   <person>
         |     <age born="1985-01-01">30</age>
         |     <name>Lars</name>
         |   </person>
         |   <person>
         |     <age born="1980-01-01">30</age>
         |     <name>Lion</name>
         |   </person>
         | </people>
      """.stripMargin

    val xmlFile = File.createTempFile("ages", ".xml")
    val writer = new BufferedWriter(new FileWriter(xmlFile))
    writer.write(xml)
    writer.close()

    println(s"Reading from $xmlFile")
    spark.stop() // stop the existing one setup in `beforeAll`
    spark = SparkSession.builder()
      .master("local[2]")
      .appName("XmlSuite")
      .config("spark.ui.enabled", false)
      .config("spark.hadoop.fs.local.block.size", "0x68")
      .getOrCreate()

    val results = spark.read
      .option("rowTag", "person")
      .option("mode", "FAILFAST")
      .format("xml")
      .load(xmlFile.getAbsolutePath)

    results.show(false)
  }
```

Setting block size to partition at "0x68" will break the second `<person>` element in the middle. Executing this code actually results in zero records being read... different symptom, but same underlying issue.

The problem seems to be inside `XmlInputFormat` where there are calls to `filePosition.getPos`. The goal seems to be to know the "index" of the character being read at any given time, so as to know when it should stop reading as it has gone beyond the split boundary

However, it seems that this call does not return the current index of the character being read, but the position in the file of the last read byte. The underlying Stream reader uses buffering (8192 bytes it seems) so the call to `filePosition.getPos` does not correspond to the index of the byte being read. It's always ahead up to 8192 bytes. So in the example above, the first call to `filePosition.getPos` will return the last byte in the file, making `readUntilStartElement` exit early and not reading any data.

In my local testing I added a mutable variable to keep track of the position, incrementing it by one after every `reader.read()` invocation, and replacing calls to `filePosition.getPos` with references to this variable. This seems to produce the desired result, both in the simple test above, as well as in the example that brought this issue to light (a 118M file that had two records dropped)

**Comment by jimenefe** [comment]
This test probably better illustrates the underlying issue. Again, inside `XmlSuite`

```scala
  test("Fails when file is partitioned in the middle of a rowTag element") {
    val xmlFile = new File(this.getClass.getClassLoader.getResource("fias_house.xml").getFile)

    spark.stop()
    spark = SparkSession.builder()
      .master("local[2]")
      .appName("XmlSuite")
      .config("spark.ui.enabled", false)
      .config("spark.hadoop.fs.local.block.size", 0x2096.toString)
      .getOrCreate()

    val results = spark.read
      .option("rowTag", "House")
      .option("mode", "FAILFAST")
      .format("xml")
      .load(xmlFile.getAbsolutePath)

    println(s"Record count : ${results.count()}")
  }
```

`0x2096` has been chosen to cause the file to be partitioned inside one of the rowTag elements

Runnint the above shows a record count of 36. If you set "spark.hadoop.fs.local.block.size" so that it doesn't break in the middle of a rowTag (for example 0x2000) then record count is 37

**Comment by jimenefe** [comment]
Cleaned the test up a bit

```scala
package com.databricks.spark.xml

import java.io.File

import org.apache.spark.sql.SparkSession
import org.scalatest.{ BeforeAndAfterAll, FunSuite, Matchers }

final class XmlPartitioningSuite extends FunSuite with Matchers with BeforeAndAfterAll {

  private var spark: SparkSession = _

  override protected def beforeAll(): Unit = {
    super.beforeAll()
    spark = SparkSession.builder()
      .master("local[2]")
      .appName("XmlPartitioningSuite")
      .config("spark.ui.enabled", false)
      .config("spark.hadoop.fs.local.block.size", 0x2096.toString) // chosen to partition input file in the middle of rowTag element 
      .getOrCreate()
  }

  override protected def afterAll(): Unit = {
    try {
      spark.stop()
      spark = null
    } finally {
      super.afterAll()
    }
  }

  test("Fails when file is partitioned in the middle of a rowTag element") {
    val xmlFile = new File(this.getClass.getClassLoader.getResource("fias_house.xml").getFile)

    val results = spark.read
      .option("rowTag", "House")
      .option("mode", "FAILFAST")
      .format("xml")
      .load(xmlFile.getAbsolutePath)

    results.count() shouldBe 37
  }
}
```

**Comment by srowen** [comment]
Yes, thank you for this. While the code does in theory handle data split across partitions (Hadoop's TextInputFormat does) and it generally works fine, I'm sure there's some weird corner issue here. It might be related to https://github.com/databricks/spark-xml/issues/338 which I never could figure out after digging pretty deep.

Thank you for the analysis so far. If you have a pull request feel free to open. @HyukjinKwon or I might be able to investigate soon too.

**Comment by jimenefe** [comment]
working on it...

**Comment by srowen** [comment]
Check out https://github.com/databricks/spark-xml/pull/400

**Comment by srowen** [comment]
This should be resolved per #400 . Thanks again!
